nowcasting_dataloader/utils/position_encoding.py

```diff
@@ -0,0 +1,291 @@
+            **kwargs,
+        )
+    # Update original dictionary
+    modalities_to_encode.update(position_encodings)
+    return modalities_to_encode
+
+
+def encode_position(
+    shape: List[int],
+    geospatial_coordinates: List[np.ndarray],
+    datetimes: List[datetime.datetime],
+    geospatial_bounds: Dict[str, float],
+    method: str = "fourier",
+    **kwargs,
+) -> torch.Tensor:
+    """
+    This function wraps a variety of different methods for generating position features for given inputs.
+
+    Args:
+        shape: The shape of the input to be encoded, should be the largest or finest-grained input
+            For example, if the inputs are shapes (12, 6, 128, 128) and (1, 6), (12, 6, 128, 128) should be passed in as
+            shape, as it has the most elements and the input (1, 6) can just subselect the position encoding
+        geospatial_coordinates: The latitude/longitude of the inputs for shape, in OSGB coordinates
+        datetimes: time of day and date for each of the timesteps in the shape
+        method: Method of the encoding, either 'fourier' for Fourier Features
+        geospatial_bounds: The bounds of the geospatial area covered, in a dict with the keys 'x_min', 'y_min', 'x_max', 'y_max'
+        kwargs: Passed to fourier_encode
+
+    Returns:
+        The position encodings for all items in the batch
+    """
+    assert method in [
+        "fourier",
+    ], AssertionError(f"method must be one of 'fourier', not '{method}'")
+
+    position_encoding = encode_absolute_position(
+        shape, geospatial_coordinates, geospatial_bounds, datetimes, **kwargs
+    )
+    return position_encoding
+
+
+def encode_absolute_position(
+    shape: List[int],
+    geospatial_coordinates: List[np.ndarray],
+    geospatial_bounds: Dict[str, float],
+    datetimes: List[datetime.datetime],
+    **kwargs,
+) -> torch.Tensor:
+    """
+    Encodes the absolute position of the pixels/voxels in time and space
+
+    This should be done per-modality and can be thought of as the relative position of the input modalities across a
+    given year and the area of the Earth covered by all the examples.
+
+    Args:
+        shape: Shape to encode positions for
+        geospatial_coordinates: The geospatial coordinates, in OSGB format
+        datetimes: Time of day and date as a list of datetimes, one for each timestep
+        geospatial_bounds: The geospatial bounds of the area where the examples come from, e.g. the coordinates of the area covered by the SEVIRI RSS image
+        **kwargs:
+
+    Returns:
+        The absolute position encoding for the given shape
+    """
+    datetime_features = create_datetime_features(datetimes)
+
+    # Fourier Features of absolute position
+    encoded_latlon = normalize_geospatial_coordinates(
+        geospatial_coordinates, geospatial_bounds, **kwargs
+    )
+
+    # Combine time and space features
+    to_concat = [einops.repeat(encoded_latlon, "b h w c -> b c t h w", t=shape[1])]
```

Comment:
I wonder if, before we have named tensors, we could use consts to identify each dimension?  So, for example, instead of doing `t=shape[1]`, instead we could do:

 ```python
TIME_DIM = 1
# ...
t=shape[TIME_DIM]
```

Maybe?

Or an Enum?

I don't know about you but I easily forget which dimension is which!

Although I don't remember seeing things like `shape[TIME_DIM]` in other peoples' PyTorch code, so maybe it's a bad idea?!